src/trace-plugin-loader.ts

```diff
@@ -16,278 +16,485 @@
+      if (instrumentation.patch) {
+        instrumentation.patch(
+            exportedValue, this.createTraceAgentInstance(logString));
+        // Queue a function to run if the plugin gets disabled.
+        if (instrumentation.unpatch) {
+          this.unpatchFns.push(() => {
+            this.logger.info(
+                `PluginWrapper#unapplyAll: [${logString}] Unpatching file.`);
+            instrumentation.unpatch!(exportedValue);
+          });
+        }
+        // The patch object should only have either patch() or intercept().
+        if (instrumentation.intercept) {
+          this.logger.warn(`PluginWrapper#applyPlugin: [${
+              logString}] Patch object has both patch() and intercept() for this file. Only applying patch().`);
+        }
+      } else if (instrumentation.intercept) {
+        exportedValue = instrumentation.intercept(
+            exportedValue, this.createTraceAgentInstance(file));
+      }
+      return exportedValue;
+    }, moduleExports as T);
   }
-  if (process._preload_modules && process._preload_modules.length > 0) {
-    const first = process._preload_modules[0];
-    if (first !== '@google-cloud/trace-agent') {
-      logger_.error(
-          'Tracing might not work as ' + first +
-          ' was loaded with --require before the trace agent was initialized.');
+
+  // Helper function to get the cached plugin value if it wasn't loaded yet.
+  getPluginExportedValue(): Plugin {
+    if (this.pluginExportedValue === ModulePluginWrapper.NOT_LOADED) {
+      this.pluginExportedValue = require(this.path);
     }
+    return this.pluginExportedValue;
   }
-}
 
-function checkPatch<T>(patch: Instrumentation<T>) {
-  if (!(patch as Patch<T>).patch && !(patch as Intercept<T>).intercept) {
-    throw new Error(
-        'Plugin for ' + patch.file + ' doesn\'t patch ' +
-        'anything.');
-  } else if ((patch as Patch<T>).patch && (patch as Intercept<T>).intercept) {
-    throw new Error(
-        'Plugin for ' + patch.file + ' has ' +
-        'both intercept and patch functions.');
-  } else if ((patch as Patch<T>).unpatch && (patch as Intercept<T>).intercept) {
-    logger_.warn(
-        'Plugin for ' + patch.file + ': unpatch is not compatible ' +
-        'with intercept.');
-  } else if ((patch as Patch<T>).patch && !(patch as Patch<T>).unpatch) {
-    logger_.warn(
-        'Plugin for ' + patch.file + ': patch method given without ' +
-        'accompanying unpatch.');
+  private createTraceAgentInstance(file: string) {
+    const traceApi = new TraceAgent(file);
+    traceApi.enable(this.logger, this.traceConfig);
+    this.traceApiInstances.push(traceApi);
+    return traceApi;
   }
 }
 
-export function activate(logger: Logger, config: PluginLoaderConfig): void {
-  if (activated) {
-    logger.error('Plugins activated more than once.');
-    return;
+/**
+ * A class that represents wrapper logic on top of plugins that patch core
```

Comment:
What does a core module mean? Does it mean a Node's builtin module?

Reply:
Yes, I'll amend the comment to that effect.